src/path.c

```diff
@@ -0,0 +1,937 @@
+
+                /* '=' */
+                LY_CHECK_GOTO(lyxp_check_token(ctx, exp, *exp_idx, LYXP_TOKEN_OPER_EQUAL), error);
+                ++(*exp_idx);
+
+                /* FuncName */
+                LY_CHECK_GOTO(lyxp_check_token(ctx, exp, *exp_idx, LYXP_TOKEN_FUNCNAME), error);
+                if ((exp->tok_len[*exp_idx] != 7) || strncmp(exp->expr + exp->tok_pos[*exp_idx], "current", 7)) {
+                    LOGVAL(ctx, LY_VLOG_NONE, NULL, LYVE_XPATH, "Invalid function \"%.*s\" invocation in path.",
+                        exp->tok_len[*exp_idx], exp->expr + exp->tok_pos[*exp_idx]);
+                    goto error;
+                }
+                ++(*exp_idx);
+
+                /* '(' */
+                LY_CHECK_GOTO(lyxp_check_token(ctx, exp, *exp_idx, LYXP_TOKEN_PAR1), error);
+                ++(*exp_idx);
+
+                /* ')' */
+                LY_CHECK_GOTO(lyxp_check_token(ctx, exp, *exp_idx, LYXP_TOKEN_PAR2), error);
+                ++(*exp_idx);
+
+                /* '/' */
+                LY_CHECK_GOTO(lyxp_check_token(ctx, exp, *exp_idx, LYXP_TOKEN_OPER_PATH), error);
+                ++(*exp_idx);
+
+                /* '..' */
+                LY_CHECK_GOTO(lyxp_check_token(ctx, exp, *exp_idx, LYXP_TOKEN_DDOT), error);
+                do {
+                    ++(*exp_idx);
+
+                    /* '/' */
+                    LY_CHECK_GOTO(lyxp_check_token(ctx, exp, *exp_idx, LYXP_TOKEN_OPER_PATH), error);
+                    ++(*exp_idx);
+                } while (!lyxp_check_token(NULL, exp, *exp_idx, LYXP_TOKEN_DDOT));
+
+                /* NameTest */
+                LY_CHECK_GOTO(lyxp_check_token(ctx, exp, *exp_idx, LYXP_TOKEN_NAMETEST), error);
+                ++(*exp_idx);
+
+                /* '/' */
+                while (!lyxp_check_token(NULL, exp, *exp_idx, LYXP_TOKEN_OPER_PATH)) {
+                    ++(*exp_idx);
+
+                    /* NameTest */
+                    LY_CHECK_GOTO(lyxp_check_token(ctx, exp, *exp_idx, LYXP_TOKEN_NAMETEST), error);
+                    ++(*exp_idx);
+                }
+            } while (!lyxp_check_token(NULL, exp, *exp_idx, LYXP_TOKEN_BRACK2)
+                    && !lyxp_check_token(NULL, exp, (*exp_idx) + 1, LYXP_TOKEN_BRACK1));
+
+        } else {
+            LOGVAL(ctx, LY_VLOG_NONE, NULL, LY_VCODE_XP_INTOK, lyxp_print_token(exp->tokens[*exp_idx]),
+                   exp->expr + exp->tok_pos[*exp_idx]);
+            goto error;
+        }
+
+        /* ']' */
+        LY_CHECK_GOTO(lyxp_check_token(ctx, exp, *exp_idx, LYXP_TOKEN_BRACK2), error);
+        ++(*exp_idx);
+    }
+
+    ly_set_free(set, NULL);
+    return LY_SUCCESS;
+
+error:
+    ly_set_free(set, NULL);
+    return LY_EVALID;
+}
+
+LY_ERR
+ly_path_parse(const struct ly_ctx *ctx, const char *str_path, size_t path_len, uint8_t begin, uint8_t lref,
+              uint8_t prefix, uint8_t pred, struct lyxp_expr **expr)
+{
+    struct lyxp_expr *exp;
+    uint16_t exp_idx;
```

Comment:
Its is called `exp_idx` because it is the index of the `exp` structure, there are several arrays there. But I suppose they both could be renamed for it to be less confusing, any ideas?

As for `lyxp_check_token()`, I do agree that "look at the next token" is not accurate (anymore), is that what you meant? Also I do not want to include index increment in the function because there are many cases (also in `xpath.c`) when it is used only to look at the next token or the token is then being processed first (such as `NameTest`) so it is not desirable to skip it immediately. Also, it would be a conditional function side effect, I would prefer to avoid those.

Reply:
ok, make sense